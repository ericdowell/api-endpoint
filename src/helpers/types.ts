import { AxiosResponse } from 'axios'

export type CatchError = <R>(error: Error, payload: QueryPayload<R>) => never | void
export type Query<R = any> = () => Promise<AxiosResponse<R>>
export interface QueryPayload<D> {
  errors: any | undefined
  message: string | undefined
  loading: boolean
  data: D | undefined
}
