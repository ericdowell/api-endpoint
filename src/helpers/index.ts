export * from './onSubmit'
export * from './safeResponseData'
export * from './types'
export * from './useQuery'
