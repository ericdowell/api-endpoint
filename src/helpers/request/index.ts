import { AxiosResponse } from 'axios'
import { RequestOptions, Request, QueryPayload } from './types'
import { processResponse } from './process'

export async function makeRequest<Data>(
  request: Request<Data>,
  payload: QueryPayload<Data>,
  options?: RequestOptions,
): Promise<void> {
  return request()
    .then((response: AxiosResponse<Data>) => processResponse(response, payload, options))
    .catch(
      async (error: Error): Promise<void> => {
        if (!options?.catchError) {
          throw error
        }
        return options.catchError(error, payload).then(() => {
          payload.loading = false
        })
      },
    )
}
