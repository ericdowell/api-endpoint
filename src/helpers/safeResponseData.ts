import { AxiosResponse } from 'axios'

export function safeResponseData<T = any>(response: AxiosResponse<T> | any, isArray = false): T {
  if (isArray) {
    return (Array.isArray(response?.data) && response.data) || []
  }
  return (response?.data && typeof response.data === 'object' && response.data) || {}
}
