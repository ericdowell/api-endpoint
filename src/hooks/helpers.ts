// TODO: Add correct types to parameters
export const applyReducerState = (state: any, key: any, action: any): any => ({
  ...state,
  [key]: action[key],
})
