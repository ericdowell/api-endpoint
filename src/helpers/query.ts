import { AxiosResponse } from 'axios'
import { safeResponseData } from './safeResponseData'
import { CatchError, Query, QueryPayload } from './types'

const DEFAULT_CATCH: CatchError = (error: Error): never => {
  throw error
}

export function makeRequest<R>(
  payload: QueryPayload<R>,
  request: Query<R>,
  options: { isArray: boolean; catchError: CatchError },
): void {
  request()
    .then((response: AxiosResponse) => {
      const { errors, message, ...data } = safeResponseData<R | any>(response, options.isArray)
      payload.errors = errors
      payload.message = message
      payload.data = data
      payload.loading = false
    })
    .catch((error) => {
      options.catchError<R>(error, payload)
      payload.loading = false
    })
}

export function query<R>(
  request: Query<R>,
  options?: { isArray?: boolean; catchError?: CatchError },
): QueryPayload<R> {
  const payload: QueryPayload<R> = {
    loading: true,
    message: undefined,
    errors: undefined,
    data: undefined,
  }
  makeRequest<R>(payload, request, {
    catchError: options?.catchError ?? DEFAULT_CATCH,
    isArray: options?.isArray ?? false,
  })
  return payload
}
