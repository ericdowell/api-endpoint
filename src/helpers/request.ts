import { AxiosResponse } from 'axios'
import { safeResponseData } from './safeResponseData'
import { CatchError, QueryOptions, Request, RequestPayload } from './types'

export const DEFAULT_REQUEST_CATCH: CatchError = (error: Error): never => {
  throw error
}

const makePayload = <Data>(options?: QueryOptions): RequestPayload<Data> => {
  return {
    data: undefined,
    errors: undefined,
    loading: options?.initialLoading ?? true,
    message: undefined,
  }
}

export function makeRequest<Data>(
  payload: RequestPayload<Data>,
  request: Request<Data>,
  options: { isArray: boolean; catchError: CatchError },
): void {
  request()
    .then((response: AxiosResponse) => {
      const { errors, message, ...data } = safeResponseData<Data | any>(response, options.isArray)
      payload.errors = errors
      payload.message = message
      payload.data = data
      payload.loading = false
    })
    .catch((error) => {
      options.catchError(error, payload)
      payload.loading = false
    })
}

export function query<Data = any>(request: Request<Data>, options?: QueryOptions): RequestPayload<Data> {
  const payload = makePayload<Data>(options)
  makeRequest<Data>(payload, request, {
    catchError: options?.catchError ?? DEFAULT_REQUEST_CATCH,
    isArray: options?.isArray ?? false,
  })
  return payload
}

export function submit<Data = any>(
  request: Request<Data>,
  options?: QueryOptions,
): RequestPayload<Data> & { onSubmit: () => Promise<void> } {
  return {
    ...makePayload<Data>({ initialLoading: false, ...options }),
    onSubmit: async function (): Promise<void> {
      this.loading = true
      await makeRequest<Data>(this, request, {
        catchError: options?.catchError ?? DEFAULT_REQUEST_CATCH,
        isArray: options?.isArray ?? false,
      })
    },
  }
}
