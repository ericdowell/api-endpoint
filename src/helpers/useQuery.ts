import { CatchError, Query, QueryPayload } from './types'
import { AxiosResponse } from 'axios'
import { safeResponseData } from './safeResponseData'

const DEFAULT_CATCH: CatchError = (error: Error): never => {
  throw error
}

export function makeRequest<R>(payload: QueryPayload<R>, request: Query<R>, catchError: CatchError): void {
  request()
    .then((response: AxiosResponse) => {
      const { errors, message, ...data } = safeResponseData<R | any>(response)
      payload.errors = errors
      payload.message = message
      payload.data = data
      payload.loading = false
    })
    .catch(catchError)
}

export function useQuery<R>(request: Query<R>, catchError: CatchError = DEFAULT_CATCH): QueryPayload<R> {
  const payload: QueryPayload<R> = {
    loading: true,
    message: undefined,
    errors: undefined,
    data: undefined,
  }
  makeRequest(payload, request, catchError)
  return payload
}
