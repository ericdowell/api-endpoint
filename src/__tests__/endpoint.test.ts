import axios, { AxiosRequestConfig } from 'axios'
import * as helper from '../helpers/safeResponseData'
import { Endpoint } from '../index'
import { BasicMock } from './mock/axios'
import qs from 'qs'

jest.spyOn(axios, 'request').mockImplementation(BasicMock)
// eslint-disable-next-line @typescript-eslint/no-empty-function
jest.spyOn(global.console, 'log').mockImplementation(() => {})

const testConfig: AxiosRequestConfig = { url: '123', method: 'get' }

describe(`${Endpoint.name}`, (): void => {
  it('the origin is localhost', () => {
    expect.assertions(1)
    expect(new Endpoint().origin).toBe('http://localhost')
  })

  it('the path is empty', () => {
    expect.assertions(1)
    expect(new Endpoint().path).toBe('')
  })

  it('the endpoint is empty', () => {
    expect.assertions(1)
    expect(new Endpoint().endpoint).toBe('')
  })

  it('the baseURL is localhost', () => {
    expect.assertions(1)
    expect(new Endpoint().baseURL).toBe('http://localhost')
  })

  it('the config.headers default is undefined', () => {
    expect.assertions(1)
    expect(new Endpoint().config.headers).toBeUndefined()
  })

  it('the paramsSerializer will return function that calls qs.stringify', () => {
    expect.assertions(2)
    const stringify = jest.spyOn(qs, 'stringify').mockImplementation((): any => 'foo')
    const paramsSerializer = new Endpoint().paramsSerializer
    expect(stringify).toHaveBeenCalledTimes(0)
    paramsSerializer({})
    expect(stringify).toHaveBeenCalledTimes(1)
    stringify.mockRestore()
  })

  it('the requestConfig will return default config', () => {
    expect.assertions(1)
    expect(new Endpoint().requestConfig({ url: 'test/path', method: 'get' })).toMatchInlineSnapshot(`
      Object {
        "baseURL": "http://localhost",
        "headers": Object {},
        "method": "get",
        "paramsSerializer": [Function],
        "url": "test/path",
      }
    `)
  })

  it('the requestConfig will return passed baseURL and paramsSerializer', () => {
    expect.assertions(1)
    const endpoint = new Endpoint()
    const config: AxiosRequestConfig = {
      baseURL: 'https://example.com',
      method: 'PATCH',
      paramsSerializer: endpoint.paramsSerializer,
      url: 'test/path',
    }
    expect(endpoint.requestConfig(config)).toMatchInlineSnapshot(`
      Object {
        "baseURL": "https://example.com",
        "headers": Object {},
        "method": "PATCH",
        "paramsSerializer": [Function],
        "url": "test/path",
      }
    `)
  })

  it('the request will return response object', async (): Promise<void> => {
    expect.assertions(1)
    const endpoint = new Endpoint()
    expect(await endpoint.request(testConfig)).toStrictEqual({
      config: {},
      data: {
        foo: 'bar',
      },
    })
  })

  it('the request will throw error axios.request throws', async (): Promise<void> => {
    expect.assertions(1)
    const error = new Error('testing')
    const endpoint = new Endpoint()
    const axiosRequest = jest.spyOn(axios, 'request').mockImplementation((): never => {
      throw error
    })
    await expect(endpoint.request(testConfig)).rejects.toBe(error)
    axiosRequest.mockRestore()
  })

  it('static method handleRequestError calls endpoint implementation of handleError', (): void => {
    expect.assertions(1)
    const error = new Error()
    expect((): void => Endpoint.handleRequestError(error, new Endpoint())).toThrow(error)
  })

  it('static method safeResponseData calls safeResponseData helper function', (): void => {
    expect.assertions(2)
    const data = undefined
    const safeResponseData = jest.spyOn(helper, 'safeResponseData').mockReturnValue(data)
    expect(Endpoint.safeResponseData({} as any)).toBe(data)
    expect(safeResponseData).toHaveBeenCalledTimes(1)
    safeResponseData.mockRestore()
  })
})
