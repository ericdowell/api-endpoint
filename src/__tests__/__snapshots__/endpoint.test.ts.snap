// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Endpoint the requestConfig will return combine headers 1`] = `
Object {
  "baseURL": "http://localhost",
  "headers": Object {
    "Accept": "text/html",
    "X-A": "baz",
    "foo": "bar",
  },
  "method": "PUT",
  "paramsSerializer": [Function],
  "url": "final/path",
}
`;

exports[`Endpoint the requestConfig will return default options 1`] = `
Object {
  "baseURL": "http://localhost",
  "method": "get",
  "paramsSerializer": [Function],
  "url": "final/path",
}
`;
